hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/OptUtils.java

```diff
@@ -343,57 +343,44 @@ public static RelDataType convert(TableField field, RelDataTypeFactory typeFacto
             return typeFactory.createTypeWithNullability(relType, true);
         }
     }
 
-    private static RelDataType convertCustomType(QueryDataType fieldType, RelDataTypeFactory typeFactory) {
-        final Map<String, RelDataType> dataTypeMap = new HashMap<>();
-        convertCustomTypeRecursively(fieldType, typeFactory, dataTypeMap);
-        return dataTypeMap.get(fieldType.getObjectTypeName());
-    }
-
-    private static void convertCustomTypeRecursively(
+    private static HazelcastObjectType convertCustomType(
             QueryDataType type,
             RelDataTypeFactory typeFactory,
-            Map<String, RelDataType> typeMap
+            Map<String, HazelcastObjectType> typeMap
     ) {
-        if (typeMap.get(type.getObjectTypeName()) != null) {
-            return;
+        HazelcastObjectType converted = typeMap.get(type.getObjectTypeName());
+        if (converted != null) {
+            return converted;
         }
 
-        final List<HazelcastObjectType.Field> fields = new ArrayList<>();
-        final HazelcastObjectTypeReference typeRef = new HazelcastObjectTypeReference();
-        typeMap.put(type.getObjectTypeName(), typeRef);
+        converted = new HazelcastObjectType(type.getObjectTypeName(), new ArrayList<>());
+        typeMap.put(type.getObjectTypeName(), converted);
 
         for (int i = 0; i < type.getObjectFields().size(); i++) {
-            final String fieldName = type.getObjectFields().get(i).getName();
-            final QueryDataType fieldType = type.getObjectFields().get(i).getDataType();
-
-            RelDataType fieldRelDataType;
-            if (fieldType.isCustomType()) {
-                fieldRelDataType = typeMap.get(fieldType.getObjectTypeName());
-                if (fieldRelDataType == null) {
-                    convertCustomTypeRecursively(fieldType, typeFactory, typeMap);
-                    fieldRelDataType = typeMap.get(fieldType.getObjectTypeName());
-                }
-            } else {
-                fieldRelDataType = typeFactory.createTypeWithNullability(
-                        typeFactory.createSqlType(HazelcastTypeUtils.toCalciteType(fieldType)),
-                        true
-                );
-            }
+            QueryDataTypeField field = type.getObjectFields().get(i);
+            QueryDataType fieldType = field.getType();
+
+            RelDataType convertedFieldType = fieldType.isCustomType()
+                    ? convertCustomType(fieldType, typeFactory, typeMap)
+                    : typeFactory.createTypeWithNullability(
+                            typeFactory.createSqlType(HazelcastTypeUtils.toCalciteType(fieldType)),
+                            true
+                    );
 
-            fields.add(new HazelcastObjectType.Field(fieldName, i, fieldRelDataType));
+            converted.getFieldList().add(new HazelcastObjectType.Field(field.getName(), i, convertedFieldType));
```

Comment:
bad encapsulation, better create method `addField` in `HazelcastObjectType`

Reply:
[`158de9fa`..`ec8c71a9`](https://github.com/hazelcast/hazelcast/compare/158de9fa4d788d55105f87227bda40a471437b39..ec8c71a9dcd6dd912fc363d2c100c1e66ddae48b)